### PR TITLE
Fix `FunctionHook` documentation

### PR DESCRIPTION
**Pull request by niboshi**
Fixed some incorrect/missing information.

### CODE REVIEW

File: chainer/function_hook.py [hunk 1]
```diff
@@ -76,15 +84,24 @@ class FunctionHook(object):
        as a thread local object. So, function hooks registered
        are different depending on threads.
 
+    If the hook is registered in this way, :meth:`~chainer.FunctionHook.added`
```
Comment (beam2d) [inline chainer/function_hook.py]: It takes time for me to notice that this sentence is a continuation of the last paragraph above the long example/note sections. How about moving it to the end of the last paragraph?

File: chainer/function_hook.py [hunk 2]
```diff
@@ -76,15 +84,24 @@ class FunctionHook(object):
...
     The other one is to register directly to
     :class:`~chainer.FunctionNode` object with
-    :meth:`~chainer.Function.add_hook` method.
+    :meth:`~chainer.FunctionNode.add_hook` method.
     Function hooks registered in this way can be removed by
-    :meth:`~chainer.Function.delete_hook` method.
+    :meth:`~chainer.FunctionNode.delete_hook` method.
     Contrary to former registration method, function hooks are registered
     only to the function which :meth:`~chainer.FunctionNode.add_hook`
     is called.
```
Comment (beam2d) [inline chainer/function_hook.py]: It's also not related to this change, but I think it's a good chance to fix some expressions:
- The other one is to `register directly` to... -> `register it directly`
- Contrary to `former registration method`, ... -> `the former registration method`
- only to the function which add_hook is called. -> only to the function whose add_hook method is called.

File: chainer/function_hook.py [hunk 3]
```diff
@@ -76,15 +84,24 @@ class FunctionHook(object):
        as a thread local object. So, function hooks registered
        are different depending on threads.
 
+    If the hook is registered in this way, :meth:`~chainer.FunctionHook.added`
+    and :meth:`~chainer.FunctionHook.deleted` are given ``None`` as the
+    ``function`` argument.
```
Comment (beam2d) [inline chainer/function_hook.py]: How about rephrasing (matching with the below one):
None is passed to the function argument of added and deleted.

File: chainer/function_hook.py [hunk 4]
```diff
@@ -76,15 +84,24 @@ class FunctionHook(object):
...
+    :meth:`~chainer.FunctionNode.add_hook` method.
     Function hooks registered in this way can be removed by
-    :meth:`~chainer.Function.delete_hook` method.
+    :meth:`~chainer.FunctionNode.delete_hook` method.
     Contrary to former registration method, function hooks are registered
     only to the function which :meth:`~chainer.FunctionNode.add_hook`
     is called.
 
+    If the hook is registered in this way, the :class:`~chainer.FunctionNode`
+    instance is given as the ``function`` argument of
```
Comment (beam2d) [inline chainer/function_hook.py]: Suggestion (similar to the above one): given as -> passed to

File: chainer/function_hook.py [hunk 5]
```diff
@@ -118,7 +136,8 @@ def deleted(self, function=None):
 
         Args:
             function(~chainer.FunctionNode): Function object to which
-                the function hook is deleted.
+                the function hook is deleted. ``None`` if the function hook
+                had been registered globally.
```
Comment (beam2d) [inline chainer/function_hook.py]: has been?

File: chainer/function_hook.py [hunk 6]
```diff
@@ -118,7 +136,8 @@ def deleted(self, function=None):
 
         Args:
             function(~chainer.FunctionNode): Function object to which
```
Comment (beam2d) [inline chainer/function_hook.py]: (again, suggestion outside of this change)
to which -> from which

File: chainer/function_hook.py (hunk 5 above)
Comment (niboshi) [reply chainer/function_hook.py]: I'm not confident about my expression either, but the hook is unregistered (the state of "being registered" is negative) at the moment of this callback.

Comment (niboshi) [reply chainer/function_hook.py]: According to @hvy, "had been" is okay but "was" is more natural expression. I fixed as such.

File: chainer/function_hook.py (hunk 3 above)
Comment (niboshi) [reply chainer/function_hook.py]: Fixed using "passed as".

File: chainer/function_hook.py (hunk 4 above)
Comment (niboshi) [reply chainer/function_hook.py]: Fixed using "passed as".

File: chainer/function_hook.py (hunk 6 above)
Comment (niboshi) [reply chainer/function_hook.py]: Fixed.

File: chainer/function_hook.py (hunk 1 above)
Comment (niboshi) [reply chainer/function_hook.py]: Fixed.

File: chainer/function_hook.py [hunk 7]
```diff
@@ -76,14 +84,24 @@ class FunctionHook(object):
        as a thread local object. So, function hooks registered
        are different depending on threads.
 
-    The other one is to register directly to
+    The other one is to register it directly to
     :class:`~chainer.FunctionNode` object with
```
Comment (hvy) [inline chainer/function_hook.py]: Nitpick but 
with -> with the

File: chainer/function_hook.py [hunk 8]
```diff
@@ -76,14 +84,24 @@ class FunctionHook(object):
...
-    :meth:`~chainer.Function.add_hook` method.
+    :meth:`~chainer.FunctionNode.add_hook` method.
     Function hooks registered in this way can be removed by
```
Comment (hvy) [inline chainer/function_hook.py]: Nitpick, actually not directly related to this PR...
registered in this way ... by -> registered this way ... by the

File: chainer/function_hook.py [hunk 9]
```diff
@@ -76,14 +84,24 @@ class FunctionHook(object):
...
-    :meth:`~chainer.Function.delete_hook` method.
-    Contrary to former registration method, function hooks are registered
-    only to the function which :meth:`~chainer.FunctionNode.add_hook`
-    is called.
+    :meth:`~chainer.FunctionNode.delete_hook` method.
+    Contrary to the former registration method, function hooks are registered
+    only to the function whose :meth:`~chainer.FunctionNode.add_hook`
+    method is called.
+
+    If the hook is registered globally using ``with`` statement, ``None`` is
```
Comment (hvy) [inline chainer/function_hook.py]: I'd place an "a" before the "``with``". What do you think?

File: chainer/function_hook.py [hunk 10]
```diff
@@ -59,9 +66,10 @@ class FunctionHook(object):
         >>> with chainer.function_hooks.TimerHook() as m:
         ...   _ = model1(x)
         ...   y = model2(x)
```
[the remark this answers is not in the document]
Comment (beam2d) [reply chainer/function_hook.py]: Yes, if you prefix the blank line with `>>>` or `...` like follows.

```py
"""
        >>> class Model(chainer.Chain):
        ...   def __init__(self):
        ...     super(Model, self).__init__()
        ...     with self.init_scope():
        ...       self.l = L.Linear(10, 10)
        ...
        ...   def __call__(self, x1):
        ...     return F.exp(self.l(x1))
        ...
        >>> model1 = Model()
        >>> model2 = Model()
        >>>
        >>> x = chainer.Variable(np.zeros((1, 10), np.float32))
        >>>
        >>> with chainer.function_hooks.TimerHook() as m:
        ...   _ = model1(x)
        ...   y = model2(x)
        ...
        >>> model3 = Model()
        >>> z = model3(y)
        >>>
        >>> print('Total time : {}'.format(m.total_time()))
        ... # doctest:+ELLIPSIS
        Total time : ...
"""
```

By the way, is it necessary to write this example in doctest style?

File: chainer/function_hook.py (hunk 3 above)
Comment (beam2d) [reply chainer/function_hook.py]: Ah, yes you are right. I was confused with "pass to function as an argument".

**Comment by toslunar** [comment]
Jenkins, test this please.

**Pull request merged**